cpp/src/arrow/compute/exec/swiss_join.cc

```diff
@@ -0,0 +1,2545 @@
+      local_states_[i].materialize.Init(pool_, proj_map_left, proj_map_right);
+    }
+
+    std::vector<JoinResultMaterialize*> materialize;
+    materialize.resize(num_threads_);
+    for (int i = 0; i < num_threads_; ++i) {
+      materialize[i] = &local_states_[i].materialize;
+    }
+
+    probe_processor_.Init(proj_map_left->num_cols(HashJoinProjection::KEY), join_type_,
+                          &hash_table_, materialize, &key_cmp_, output_batch_callback_);
+
+    InitTaskGroups();
+
+    return Status::OK();
+  }
+
+  void InitTaskGroups() {
+    task_group_build_ = scheduler_->RegisterTaskGroup(
+        [this](size_t thread_index, int64_t task_id) -> Status {
+          return BuildTask(thread_index, task_id);
+        },
+        [this](size_t thread_index) -> Status { return BuildFinished(thread_index); });
+    task_group_merge_ = scheduler_->RegisterTaskGroup(
+        [this](size_t thread_index, int64_t task_id) -> Status {
+          return MergeTask(thread_index, task_id);
+        },
+        [this](size_t thread_index) -> Status { return MergeFinished(thread_index); });
+    task_group_scan_ = scheduler_->RegisterTaskGroup(
+        [this](size_t thread_index, int64_t task_id) -> Status {
+          return ScanTask(thread_index, task_id);
+        },
+        [this](size_t thread_index) -> Status { return ScanFinished(thread_index); });
+  }
+
+  Status ProbeSingleBatch(size_t thread_index, ExecBatch batch) override {
+    if (IsCancelled()) {
+      return status();
+    }
+
+    if (!local_states_[thread_index].hash_table_ready) {
+      local_states_[thread_index].hash_table_ready = hash_table_ready_.load();
+    }
+    ARROW_DCHECK(local_states_[thread_index].hash_table_ready);
+
+    ExecBatch keypayload_batch;
+    ARROW_ASSIGN_OR_RAISE(keypayload_batch, KeyPayloadFromInput(/*side=*/0, &batch));
+
+    return CancelIfNotOK(probe_processor_.OnNextBatch(
+        thread_index, keypayload_batch, &local_states_[thread_index].temp_stack,
+        &local_states_[thread_index].temp_column_arrays));
+  }
+
+  Status ProbingFinished(size_t thread_index) override {
+    if (IsCancelled()) {
+      return status();
+    }
+
+    return CancelIfNotOK(ScanHashTableAsync(static_cast<int64_t>(thread_index)));
+  }
+
+  Status BuildHashTable(size_t thread_id, AccumulationQueue batches,
+                        BuildFinishedCallback on_finished) override {
+    if (IsCancelled()) {
+      return status();
+    }
+
+    build_side_batches_ = std::move(batches);
+    build_finished_callback_ = on_finished;
+
+    return CancelIfNotOK(BuildHashTableAsync(static_cast<int64_t>(thread_id)));
+  }
+
+  void Abort(TaskScheduler::AbortContinuationImpl pos_abort_callback) override {
+    EVENT(span_, "Abort");
+    END_SPAN(span_);
+    std::ignore = CancelIfNotOK(Status::Cancelled("Hash Join Cancelled"));
+    scheduler_->Abort(std::move(pos_abort_callback));
+  }
+
+ private:
+  Status BuildHashTableAsync(int64_t thread_id) {
```

Comment:
Logically it is still an asynchronous operation. It does not use a future, but it does use a callback function (fixed callback function) when the operation is finished. So the fact that an operation like here is a) potentially parallel and b) asynchronous, seems important enough to have some kind of common suffix for it. Because there will be a lot more of this kind of functions and they will use the same pattern with the callback that may not be obvious right away when reading the code.

But for now, I can rename them.